addons/mail/models/discuss/ir_binary.py

```diff
@@ -21,4 +21,27 @@ def _find_record_check_access(self, record, access_token, field):
                     domain.append(("channel_member_ids", "any", [("guest_id", "=", current_guest.id)]))
                 if self.env["discuss.channel"].search_count(domain, limit=1):
                     return record.sudo()
+                if current_guest:
+                    # access to the record is allowed if there are common channels
+                    # between record as a guest or partner and current guest
+                    if record._name == "res.partner":
+                        domain = [("author_id", "=", record.id), ("model", "=", "discuss.channel")]
+                    else:
+                        domain = [
+                            ("author_guest_id", "=", record.id),
+                            ("model", "=", "discuss.channel"),
```

Comment:
I don't think there is a reason to limit to to channels, what we want to check is (the thread of) the message is accessible.

This should be built in message acl check included in _search, if the method is not called as sudo (to verified).